container-search/src/main/java/com/yahoo/prelude/fastsearch/FS4InvokerFactory.java

```diff
@@ -0,0 +1,115 @@
+ * content nodes or dispatchers over the fnet/FS4 protocol
+ *
+ * @author ollivir
+ */
+public class FS4InvokerFactory {
+    private final FS4ResourcePool fs4ResourcePool;
+    private final VespaBackEndSearcher searcher;
+    private final ImmutableMap<Integer, SearchCluster.Node> nodesByKey;
+
+    public FS4InvokerFactory(FS4ResourcePool fs4ResourcePool, SearchCluster searchCluster, VespaBackEndSearcher searcher) {
+        this.fs4ResourcePool = fs4ResourcePool;
+        this.searcher = searcher;
+
+        ImmutableMap.Builder<Integer, SearchCluster.Node> builder = ImmutableMap.builder();
+        searchCluster.groups().values().forEach(group -> group.nodes().forEach(node -> builder.put(node.key(), node)));
+        this.nodesByKey = builder.build();
+    }
+
+    public SearchInvoker getSearchInvoker(Query query, SearchCluster.Node node) {
+        return new FS4SearchInvoker(searcher, query, fs4ResourcePool, node.hostname(), node.fs4port(), node.key());
+    }
+
+    public Optional<SearchInvoker> getSearchInvoker(Query query, SearchCluster.Group group) {
+        return getInvoker(group.nodes(), node -> getSearchInvoker(query, node), InterleavedSearchInvoker::new);
+    }
+
+    public FillInvoker getFillInvoker(Query query, SearchCluster.Node node) {
+        return new FS4FillInvoker(searcher, query, fs4ResourcePool, node.hostname(), node.fs4port(), node.key());
+    }
+
+    public Optional<FillInvoker> getFillInvoker(Result result) {
+        Collection<Integer> requiredNodes = requiredFillNodes(result);
+        List<SearchCluster.Node> nodes = new ArrayList<>(requiredNodes.size());
+
+        for (Integer distKey : requiredNodes) {
+            SearchCluster.Node node = nodesByKey.get(distKey);
+            if (node == null) {
+                return Optional.empty();
+            }
+            nodes.add(node);
+        }
+
+        Query query = result.getQuery();
+        return getInvoker(nodes, node -> getFillInvoker(query, node), InterleavedFillInvoker::new);
+    }
+
+    private static Collection<Integer> requiredFillNodes(Result result) {
+        Set<Integer> requiredNodes = new HashSet<>();
+        for (Iterator<Hit> i = result.hits().unorderedDeepIterator(); i.hasNext();) {
+            Hit h = i.next();
+            if (h instanceof FastHit) {
+                FastHit hit = (FastHit) h;
+                requiredNodes.add(hit.getDistributionKey());
+            }
+        }
+        return requiredNodes;
+    }
+
+    @FunctionalInterface
+    private interface InvokerConstructor<INVOKER> {
+        INVOKER construct(SearchCluster.Node node);
+    }
+
+    @FunctionalInterface
+    private interface ClusterInvokerConstructor<CLUSTERINVOKER extends INVOKER, INVOKER> {
+        CLUSTERINVOKER construct(Map<Integer, INVOKER> subinvokers);
+    }
+
+    private <INVOKER extends CloseableInvoker, CLUSTERINVOKER extends INVOKER> Optional<INVOKER> getInvoker(
```

Comment:
This is unusual and hard to understand - could you leave an explanation for future confused individuals?

Reply:
Yeah, it's an attempt to avoid duplicating instantiation code when the two invoker types don't share much in the way of class hierarchy. I don't know if the lambda magic is really worth the different kind of complexity, but I'll leave it in for now and add comments to it in the next iteration.